Add Config Flow to bmw_connected_drive

rikroe opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
BMW Connected Drive has been migrated to config flows. Your old entries from configuration.yaml are used __once__ for initial setup. It is recommended to rename the device_tracker of your car(s) in config/known_devices.yaml (you can e.g. add `_old` to the entity name) before updating, otherwise a second device tracker entity (ending with `_2`) per car will be created. After updating, the entity can be removed from known_devices as that will no longer be used.


## Proposed change
This PR adds a config flow to `bmw_connected_drive`. The existing `configuration.yaml` is loaded initially. The device tracker update was required to allow for config flows. Non-mandatory options can be configured from an option flow.
All platforms now only require to distinguish between the different vehicles of an account, as multiple accounts are stored in different config entries. 

This PR also adds information to the device registry where possible. To make it easy to distighuish between cars, I am using the vehicle identification number (VIN) as firmware (due to the lack of another displayed field in the device registry).

As suggested by @gerard33, I am adding myself as code owner.

Not in scope of this PR (but planned): Creating a `BMWConnectedDriveBaseEntity` to be used in all platforms in combination with `DataUpdateCoordinator`.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

Existing entries are imported once and can be deleted afterwards in favor of config flows.

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #29976 due to rewritten device tracker
- This PR is related to issue: includes a fix from #14464 for older vehicles
- Link to documentation pull request: home-assistant/home-assistant.io#13471

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository] home-assistant/home-assistant.io#13471

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

rikroe committed:
Add Config Flow to bmw_connected_drive

rikroe committed:
Fix test requirements for bmw_connected_drive

rikroe committed:
Create option flow update listener only once

gerard33 reviewed (approved):
LGTM

dgomes commented on homeassistant/components/bmw_connected_drive/lock.py:
```diff
@@ -42,6 +41,17 @@ def __init__(self, account, vehicle, attribute: str, sensor_name):
         )
 
+    @property
+    def device_info(self) -> dict:
```
Can we find a way of sharing this function between entities ?

rikroe replied:
It definitely can when using something like a `BMWBaseEntity`.
I left that out as I have another PR with `DataUpdateCoordinator` where this is part and I didn't want to increase the size of this PR even more.

dgomes commented on homeassistant/components/bmw_connected_drive/translations/en.json:
```diff
@@ -0,0 +1,35 @@
+{
```
Unsure if en.json can't be generated from string.json

rikroe replied:
It is generated from string.json using `python3 -m script.translations develop` to see some text in the dev environment. 
I just commited it with the rest as there was no  gitignore etc. but can remove if needed.

dgomes reviewed (commented):
Looks good

rikroe commented:
@frenck sorry to reference you directly, but as you approved the docs PR... Any chance this can get reviewed by the HA core team?
I know you have lots to do but would appreciate any hints. Also fine with totally refactoring this PR although I'm not really sure how (especially how to split it up).
Thanks!

frenck committed:
Bumped version to 0.114.0b0

MartinHjelmare committed:
Fix missing rfxtrx strings (#38570)

* Fix missing rfxtrx strings

* Clean

pvizeli committed:
Revert "Add a timeout for async_add_entities (#38474)" (#38584)

This reverts commit 7590af393077fbee840a7e91921717a4b699a553.

emontnemery committed:
Suppress MQTT discovery updates without changes (#38568)

Quentame committed:
Remove Linky integration (#38565)

balloob committed:
Do not print warning when command line switch queries off (#38591)

starkillerOG committed:
Improve Xioami Aqara zeroconf discovery handling (#37469)

Co-authored-by: Franck Nijhof <git@frenck.dev>
Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

balloob committed:
Bumped version to 0.114.0b1

timmo001 committed:
Improve the OVO Energy integration (#38598)

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

adrum committed:
Expose video doorbell button state to HomeKit (#38617)

bdraco committed:
Ensure homekit pairing barcode is usable on dark themes (#38609)

balloob committed:
Handle unavailable input_select in Google Assistant (#38611)

pvizeli committed:
V2 timeout for async_add_entities (#38601)

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>
Co-authored-by: J. Nick Koston <nick@koston.org>
Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

pvizeli committed:
Bump OpenCV 4.3.0 and Numpy 1.19.1 (#38616)

bdraco committed:
Ensure doorbird does not block startup (#38619)

hunterjm committed:
Upgrade to TensorFlow 2 (#38384)

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>
Co-authored-by: Martin Hjelmare <marhje52@gmail.com>
Co-authored-by: Franck Nijhof <git@frenck.dev>

balloob committed:
Fix lint

balloob committed:
Bumped version to 0.114.0b2

frenck committed:
Remove tf-models-official from wheels builder (#38637)

Swampen committed:
Fix url in seventeentrack delivered notification (#38646)

bachya committed:
Bump regenmaschine to 2.1.0 (#38649)

ctalkington committed:
Fix AccuWeather async timeout (#38654)

AlejandroRivera committed:
Fix rest_command UnboundLocalError in exception handling (#38656)

```
2020-08-07 22:38:10 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection.3903193064] local variable 'response' referenced before assignment
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/rest_command/__init__.py", line 115, in async_service_handler
    async with getattr(websession, method)(
  File "/usr/local/lib/python3.8/site-packages/aiohttp/client.py", line 1012, in __aenter__
    self._resp = await self._coro
  File "/usr/local/lib/python3.8/site-packages/aiohttp/client.py", line 582, in _request
    break
  File "/usr/local/lib/python3.8/site-packages/aiohttp/helpers.py", line 586, in __exit__
    raise asyncio.TimeoutError from None
asyncio.exceptions.TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 125, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1281, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1316, in _execute_service
    await handler.func(service_call)
  File "/usr/src/homeassistant/homeassistant/components/rest_command/__init__.py", line 137, in async_service_handler
    _LOGGER.warning("Timeout call %s", response.url, exc_info=1)
UnboundLocalError: local variable 'response' referenced before assignment
```

starkillerOG committed:
Fix xiaomi_aqara discovery (#38622)

ludeeus committed:
Update frontend to 20200807.1 (#38626)

bdraco committed:
Ensure shared zeroconf is passed to homekit controller devices (#38678)

Tho85 committed:
Make sure groups are initialized before template sensors (#37766)

* Make sure groups are initialized before template sensors

This way users may use the `expand` function in templates to expand
groups and have HA listen for changes to group members.

Fixes #35872

* Patch async_setup_platform instead of async_setup

* Cleanup

* Use an event to avoid sleep

* Update tests/components/template/test_sensor.py

Co-authored-by: J. Nick Koston <nick@koston.org>

Co-authored-by: J. Nick Koston <nick@koston.org>

bachya committed:
Fix missing data for Guardian "AP enabled" binary sensor (#38681)

OnFreund committed:
Bump pyvolumio to 0.1.1 (#38685)

bdraco committed:
Update aiohomekit to handle homekit devices that do not send format (#38679)

balloob committed:
Bumped version to 0.114.0b3

balloob committed:
Bump updater timeout (#38690)

pnbruckner committed:
Fix async_fire_time_changed in tests/common.py (#38705)

pvizeli committed:
Update base image 8.2.1 (#38716)

balloob committed:
Bumped version to 0.114.0b4

pvizeli committed:
Add scikit-build to installed env (#38726)

balloob committed:
Add scan_tag webhook to mobile app (#38721)

andrewsayre committed:
Bump pysmartthings 0.7.3 (#38732)

dmulcahey committed:
Make default duration 1/10th of a second for ZHA light calls (#38739)

* default duration to 1/10th of a second
* update test

bdraco committed:
Install a threading.excepthook on python 3.8 and later (#38741)

Exceptions in threads were being silently discarded and never
logged as the new in python 3.8 threading.excepthook was not
being set.

etheralm committed:
Bump dyson upstream library version (#38756)

ludeeus committed:
Bump frontend to 20200811.0 (#38760)

dmulcahey committed:
Bump ZHA quirks lib to 0.0.43 (#38762)

Adminiuga committed:
Bump up ZHA dependencies (#38775)

frenck committed:
Fix lastest version in updater for Supervisor enabled installs (#38773)

* Fix lastest version in update for Supervisor enabled installs

* Fix updater tests

frenck committed:
Bumped version to 0.114.0

frenck committed:
Merge pull request #38785 from home-assistant/rc

bachya committed:
Handle unhandled exceptions related to unavailable SimpliSafe features (#38812)

bdraco committed:
Make executor max_workers consistent between python versions (#38821)

The default on python 3.8 is for max_workers is significantly
lower than the default on python 3.7 which means we can get starved
for workers.

To determine a reasonable maximum, the maximum was increased to large
number on 5 production instances.

The number of worker threads created during startup that were
needed to avoid waiting for a thread:

  HOU 1 - 71
  HOU 2 - 48
  OGG 1 - 60
  OGG 2 - 68
  OGG 3 - 64

This lead to a selection of 64 as it was reliable in all cases
and did not have a significant memory impact

RogerSelwyn committed:
Fix creation of unrequired sensors in OVO energy (#38835)

balloob committed:
Catch upnp timeout error (#38794)

esev committed:
Bump pywemo to 0.4.46 (#38845)

bdraco committed:
Ensure service browser does not collapse on bad dns names (#38851)

If a device on the network presented a bad name, zeroconf
would throw zeroconf.BadTypeInNameException and the service
browser would die off.  We now trap the exception and continue.

timmo001 committed:
Fix OVO Energy Sensors (#38849)

firstof9 committed:
Fix ozw dimming transition (#38850)

* Handle float from light component

* Test with float

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

Quentame committed:
Update meteo_france based on code review (#38789)

* Review: if not to pop

* Review: async_add_job --> async_add_executor_job

* Review: const

* Review: start logging messages with capital letter

* Review : UTC isoformated time --> fix "Invalid date""

* Fix hail forecast condition

* Review: _show_setup_form is a callback

* Fix update option

* Review: no icon for next_rain

* Review: inline cities form

* Review: store places as an instance attribute

* UNDO_UPDATE_LISTENER()

timmo001 committed:
Update ovoenergy package to v1.1.7 (#38875)

bdraco committed:
Adjust slow add entities timeouts to handle slowest known case (#38876)

With this change, we should still be able to startup
in under 10 minutes if something really goes wrong.

The testing done in #38661 was used to determine
these values.

balloob committed:
Bumped version to 0.114.1

balloob committed:
Merge pull request #38894 from home-assistant/rc

oncleben31 committed:
Fix error in meteo_france for overseas France cities (#38895)

escoand committed:
Better timeout handling in samsungtv integration (#38759)

* handle PlatformNotReady

* set timeout in bridge

* set timeout in test

* Revert "handle PlatformNotReady"

This reverts commit 118ee06ba016f88abdf767efdf4ef1eacb4e5caa.

MartinHjelmare committed:
Fix ozw pure rgb dimmer light (#38877)

* Fix ozw pure rgb light

* Add test

cgtobi committed:
Fix Netatmo climate boost/heat event handling (#38923)

* Fix boost event handling

* Replace strings with vars

automaton82 committed:
Fix the CONF_LOOP check to use the config (#38890)

tizzen33 committed:
Fix 'Not Available' message for Onkyo integration (#38554)

pbalogh77 committed:
Fix HC3 compatibility further (#38931)

* Update __init__.py

Further fixes for HC3 compatibility.

* Update homeassistant/components/fibaro/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Bump netdisco to 2.8.2 to accomodate new zeroconf exception (#38949)

bdraco committed:
Accommodate systems with very large databases and slow disk/cpu (#38947)

On startup we run an sqlite3 quick_check to verify the database
integrity. In the majority of cases, the quick_check takes under
10 seconds.

On systems with very large databases and very slow disk/cpu,
this can take much longer so we freeze the timeout.

balloob committed:
Bumped version to 0.114.2

balloob committed:
Merge pull request #38964 from home-assistant/rc

bdraco committed:
Update zeroconf to fix ServiceBrowser leak on cancelation (#38933)

emontnemery committed:
Bump pychromecast to 7.2.1 (#39018)

lawtancool committed:
Fix Control4 light setup issues (#38952)

* catch KeyError and UnboundLocalError in light setup

* add check for no dimmer info from control4

* use existence of light_level data to enable dimming

* check if light data is not in dimmer and non_dimmer

* add item_variables logging

* add await to item_variables call

* add full item dump

* remove full item list

* change logging message

* fix typo

* reduce code inside try and add all entities at once

* Apply suggestions from code review

Co-authored-by: J. Nick Koston <nick@koston.org>

* format with black

Co-authored-by: J. Nick Koston <nick@koston.org>

Danielhiversen committed:
Update met.no library (#39076)

bdraco committed:
Fix emulated hue on/off devices compatibility with alexa (#39063)

balloob committed:
Bumped version to 0.114.3

balloob committed:
Merge pull request #39079 from home-assistant/rc

pvizeli committed:
Fix TTS languange characters (#39211)

bdraco committed:
Fix time pattern listener firing a few microseconds early (#39281)

balloob committed:
Bumped version to 0.114.4

balloob committed:
Merge pull request #39286 from home-assistant/rc

dgomes commented:
can we fix the conflict ?

rikroe commented:
Woah, rebased on `master` instead of `dev`... Will try to see if I'm able to fix that...

rikroe commented:
Completely bricked this one, opened up #39585 instead and closing this.
